unix/w0vncserver/PipeWirePixelBuffer.cxx

```diff
@@ -0,0 +1,259 @@
+{
+  setSize(width, height);
+  setPF(pf);
+  server->setPixelBuffer(this);
+}
+
+void PipeWirePixelBuffer::processBuffer(pw_buffer* buffer)
+{
+  pixman_bool_t ret;
+  spa_buffer* buf;
+  spa_meta_region* damage;
+  int srcStride;
+  int dstStride;
+  uint8_t* srcBuffer;
+  uint8_t* dstBuffer;
+
+  buf = buffer->buffer;
+  srcBuffer = (uint8_t*)buf->datas[0].data;
+  srcStride = buf->datas[0].chunk->stride / (getPF().bpp / 8);
+  dstBuffer = getBufferRW({0, 0, width(), height()}, &dstStride);
+
+  ret = pixman_blt((uint32_t*)srcBuffer, (uint32_t*)dstBuffer,
+                   srcStride, dstStride, getPF().bpp,
+                   getPF().bpp, 0, 0, 0, 0, width(), height());
+
+  if (!ret)
+    imageRect(getPF(), {0, 0, width(), height()}, srcBuffer, srcStride);
+
+  server->add_changed({{0, 0, width(), height()}});
+  damage = (spa_meta_region *)spa_buffer_find_meta_data(buf,
+                                                        SPA_META_VideoDamage,
+                                                        sizeof(*damage));
+  if (damage) {
+    core::Point tl{damage->region.position.x, damage->region.position.y};
+    core::Point br{static_cast<int>(damage->region.position.x + damage->region.size.width),
+                   static_cast<int>(damage->region.position.y + damage->region.size.height)};
+    server->add_changed({{tl, br}});
+  } else {
+    server->add_changed({{0, 0, width(), height()}});
+  }
+}
+
+void PipeWirePixelBuffer::processCursor(pw_buffer* buf)
+{
+  spa_meta_cursor* cursorData;
+  spa_meta_bitmap* cursorBitmap;
+  uint8_t* buffer;
+
+  cursorData = (spa_meta_cursor *)spa_buffer_find_meta_data(buf->buffer,
+                                                            SPA_META_Cursor,
+                                                            sizeof(*cursorData));
+  if(!cursorData){
+    vlog.error("Could not find cursor metadata");
+    return;
+  }
+
+  cursor->x = cursorData->position.x;
+  cursor->y = cursorData->position.y;
+  cursor->hotspotX = cursorData->hotspot.x;
+  cursor->hotspotY = cursorData->hotspot.y;
+
+  server->setCursorPos({cursor->x, cursor->y}, true);
+
+  // No new cursor bitmap
+  if (!cursorData->bitmap_offset)
+    return;
+
+  cursorBitmap = SPA_PTROFF(cursorData, cursorData->bitmap_offset,
+                            struct spa_meta_bitmap);
+  if (!supportedCursorPixelformat(cursorBitmap->format))
+    return;
+
+  buffer = SPA_PTROFF(cursorBitmap, cursorBitmap->offset, uint8_t);
+
+  cursor->w = cursorBitmap->size.width;
+  cursor->h = cursorBitmap->size.height;
+  cursor->stride = cursorBitmap->stride;
+
+  setCursor(cursor->w, cursor->h, cursor->hotspotX,
+            cursor->hotspotY, buffer);
+}
+
+rfb::PixelFormat PipeWirePixelBuffer::convertPixelformat(int spaFormat)
+{
+  switch (spaFormat) {
+  case SPA_VIDEO_FORMAT_BGRx:
+  case SPA_VIDEO_FORMAT_BGRA:
+    return rfb::PixelFormat(32, 24, true, true, 255, 255, 255,
+                            8, 16, 24);
```

Comment:
This seems incorrect. You shouldn't toggle endianness if you've also changed the bit shifts. They'll cancel each other out.